src/liballoc/collections/btree/map.rs

```diff
@@ -626,11 +637,14 @@ impl<K: Ord, V> BTreeMap<K, V> {
     /// ```
     #[unstable(feature = "map_first_last", issue = "62924")]
     pub fn first_entry(&mut self) -> Option<OccupiedEntry<'_, K, V>> {
-        let front = self.root.as_mut()?.as_mut().first_leaf_edge();
+        let root = self.root.as_mut()?;
+        let root_ptr = unsafe { NonNull::new_unchecked(root as *mut _) };
+        let front = root.as_mut().first_leaf_edge();
```

Comment:
Hm, so actually I think there may be some problem here -- but one that we could fix relatively easily. For example, here, we first derive the raw pointer (`root_ptr`) from `root`, and then go on to make use of `root` again. I think this may be what @CAD97 is addressing when they say we're invalidating the raw pointer.

I think this could should be correct, though, if we derive the `front` pointer (and indeed anything from the root) from the unique NonNull. e.g., `let front = root_ptr.as_mut().first_leaf_edge();`. That way we're never forming a unique reference without going through the raw pointer.

I believe there should be *some* pattern (though perhaps not under Stacked Borrows, I'm not sure -- @RalfJung would know best) that allows us to stash away a raw pointer to the root and then access it at the end.

Reply:
Yeah, that's what I've pretty much arrived at now. In fact, I just realized I already fought the same enemy in `drain_filter_inner`, ended up with a trick to map root in two things. Now this becomes a little more elegant.

Reply:
I'm pushing a commit with this that passes fine on Miri, just to share, but I still want to add some comments. And I'm tempted to rename `Root::as_ref` and such, because they don't convert/expose the root, but only part of it.

Reply:
> I believe there should be some pattern (though perhaps not under Stacked Borrows, I'm not sure -- @RalfJung would know best) that allows us to stash away a raw pointer to the root and then access it at the end.

Yes -- don't use its ancestor `&mut` again, ever. Use only raw pointers anywhere that could alias with this raw root pointers.

Reply:
I think that's pretty securely enforced in the latest code. Though later some other code proved I still don't understand borrowing properly.